### PR TITLE
feat: handle exceptions; improve logs

### PR DESCRIPTION
**Pull request by weizihong**

### CODE REVIEW

File: server.hpp [hunk 1]
```diff
@@ -160,33 +188,42 @@ class Room {
...
-                // TODO: support multiple remote waiting players
-                deliver_to_others(msg, participant);
-            } else {
-                Player player { participant, name, role, PlayerType::REMOTE_HUMAN_PLAYER };
-                contest.enroll(std::move(player));
-                deliver_to_local(msg);
+            Player player;
+            try {
+                if (participant->is_local) {
+                    player = Player { participant, name, role, PlayerType::LOCAL_HUMAN_PLAYER };
```
Comment (andylizf) [inline server.hpp:209]: Split this huge code blick. You can first determine the player type. like 
```c++
auto player_type { participant->is_local ? PlayereType::LOCAL_HUMAN_PLAYER : PlayerType::REMOTE_HUMAN_PLAYER};
```
And construct the player.
Then, use the try-catch block and enroll the player.
Finally, deliver_to_what based on is_local

**Review by andylizf — CHANGES_REQUESTED** [review]

File: contest.hpp [hunk 2]
```diff
@@ -117,30 +134,22 @@ class PlayerList {
...
-            }
-        }
-        if (contains(player.role)) {
-            logger->critical("Insert player: {} role already occupied", player.role.to_string());
-            throw std::logic_error("Role already occupied");
+            } else
+                throw PlayerListFull("No role for player");
         }
+        if (contains(player.role))
+            throw RoleOccupied("Role already occupied");
```
[the remark this answers is not in the document]
Comment (weizihong) [reply contest.hpp:150]: I don't understand what should I do.

File: contest.hpp [hunk 3]
```diff
@@ -172,6 +181,9 @@ _EXPORT class Contest {
     class TimeLimitExceededException : public std::runtime_error {
         using runtime_error::runtime_error;
     };
+    class StatusError : public std::logic_error {
```
[the remark this answers is not in the document]
Comment (weizihong) [reply contest.hpp:184]: Enroll or reject player also use StatusError.

File: contest.hpp [hunk 4]
```diff
@@ -227,8 +224,11 @@ _EXPORT class Contest {
             throw StatusError("Contest not started");
         if (current.role != player.role)
             throw std::logic_error(player.name + " not allowed to play");
-        if (current.board[pos])
-            throw StonePositionitionOccupiedException("Stone positionition occupied");
+        if (current.board[pos]) {
+            status = Status::GAME_OVER;
+            result = { -player.role, WinType::SUICIDE };
```
Comment (andylizf) [inline contest.hpp:229]: logger.warn

File: server.hpp [hunk 5]
```diff
@@ -320,10 +311,10 @@ class Room {
             try {
                 contest.concede(player);
             } catch (Contest::StatusError& e) {
-                logger->error("Give up: Contest stautus is {}", std::to_underlying(contest.status));
+                logger->error("Concede: Contest stautus is {}", std::to_underlying(contest.status));
```
[the remark this answers is not in the document]
Comment (andylizf) [reply server.hpp:314]: BTW, typo. It's `status`

File: contest.hpp [hunk 6]
```diff
@@ -135,7 +135,7 @@ class PlayerList {
     auto insert(Player&& player)
     {
         if (std::ranges::find(players, player) != players.end())
-            throw PlayerExist("Player already in list");
+            throw PlayerExistException("Player already in list");
```
Comment (andylizf) [inline contest.hpp:138]: use {} to construct

File: contest.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (andylizf) [reply contest.hpp]: like `throw RoleOccupiedException{ "Role already occupied" };`

**Review by andylizf — APPROVED** [review]

**Pull request merged**